hasher-matcher-actioner/hmalib/lambdas/api/api_root.py

```diff
@@ -71,12 +78,72 @@ def root():
 
 def lambda_handler(event, context):
     """
-    root request handler
+    This lambda is invoked in 2 situations:
+
+    1. When the API is called, it uses bottle to process the request and send it to the direct function
+
+    2. Platforms can connect their AWS S3 Buckets directly to HMA so that uploads to those buckets are
+    fed directly into the system. When an upload occurs, this lambda is invoked with an s3 event. We then
+    convert the event into a URL which we submit to the hasher (via SNS)
     """
+    if is_s3_event(event):
+        logger.info(
+            "Lambda triggered with S3 event. Converting to submit content request."
+        )
+        for record in event["Records"]:
+            record = record["s3"]
+            if record["object"]["size"] == 0:
+                # ignore folders and empty files
+                continue
+            response = submit_from_url(
+                submit_content_request_from_s3_event_record(record),
+                dynamodb_table=dynamodb.Table(DYNAMODB_TABLE),
+                images_topic_arn=IMAGES_TOPIC_ARN,
+            )
+            if (
+                isinstance(response, SubmitContentResponse)
+                and response.submit_successful
+            ):
+                logger.info(f"Sucessfully submitted s3 event record as url upload.")
+            else:
+                logger.info(
+                    "Failed to submit s3 event record as url upload. "
+                    + response.message
+                    if isinstance(response, SubmitContentError)
```

Comment:
We are swallowing the error cases here. That would prevent the queue from doing retries. Shouldn't we raise here instead?

If you are repurposing an HTTP handler or view for another objective, you'll keep converting between the view classes and your events.

Instead, extract the meat of submit_from_url into a method that raises exceptions when failed. 
```
def submit_from_url_or_a_better_name(url, queue, dynamodb) 
# use primitive arguments rather than a serializable. Serializability is helpful
# when doing remote calls or in HTTP, usually limiting when doing python - 
# python calls.
# This can raise exceptions. Or do pretty much anything the python way.

@submit_api.route
def submit_from_url_api_method():
  try: submit_from_url() except Exception e: return SubmitContentResponse
```

Keep the HTTP ness of the method hidden from the s3 event handing portions.

Reply:
discussed offline. `submit_from_url` can never return an error (only throw exceptions) so simplifying logic